tsuru/client/plugin.go

```diff
@@ -57,37 +62,208 @@ func (c *PluginInstall) Run(context *cmd.Context, client *cmd.Client) error {
 	pluginName := context.Args[0]
 	pluginURL := context.Args[1]
 	if err := installPlugin(pluginName, pluginURL); err != nil {
-		return err
+		return fmt.Errorf("Error installing plugin %q: %w", pluginName, err)
 	}
 
 	fmt.Fprintf(context.Stdout, `Plugin "%s" successfully installed!`+"\n", pluginName)
 	return nil
 }
 
 func installPlugin(pluginName, pluginURL string) error {
-	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
-	file, err := filesystem().OpenFile(pluginPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+	tmpDir, err := filesystem().MkdirTemp("", "")
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not create a tmpdir: %w", err)
 	}
+	defer filesystem().RemoveAll(tmpDir)
+
 	resp, err := http.Get(pluginURL)
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not GET %q: %w", pluginURL, err)
 	}
+
 	defer resp.Body.Close()
-	data, err := ioutil.ReadAll(resp.Body)
+
+	data, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return err
 	}
 	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
 		return fmt.Errorf("Invalid status code reading plugin: %d - %q", resp.StatusCode, string(data))
 	}
-	n, err := file.Write(data)
+
+	// Try to extract .tar.gz first, then .zip. Fallbacks to copy the content
+	extractErr := extractTarGz(tmpDir, bytes.NewReader(data))
+	if extractErr != nil {
+		extractErr = extractZip(tmpDir, bytes.NewReader(data))
+	}
+	if extractErr != nil {
+		file, err := filesystem().OpenFile(filepath.Join(tmpDir, pluginName), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+		if err != nil {
+			return fmt.Errorf("Failed to open file: %w", err)
+		}
+		defer file.Close()
+		n, err := file.Write(data)
+		if err != nil {
+			return fmt.Errorf("Failed to write file content: %w", err)
+		}
+		if n != len(data) {
+			return fmt.Errorf("Incomplete write")
+		}
+	}
+
+	executablePath := findExecutablePlugin(tmpDir, pluginName)
+	if executablePath == "" {
+		return fmt.Errorf("The downloaded plugin content is invalid.")
+	}
+
+	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
+	if extractErr == nil {
+		os.Chmod(tmpDir, 0755)
```

Comment:
```suggestion
		os.Chmod(executablePath, 0755)
```

Shouldn't it be just in the executable file instead?

Reply:
The directory created by os.MkdirTemp() has `0700` which is not the default on most OS.
But I think doing the Chmod after moving the tmp dir is clearer, I will move it

The most correct way would be to check the `umask` value, but that is unix specific, which would not compile easily into windows.
On windows the Chmod would soft fail (and that is ok)

Reply:
🤔 If the compacted file was built correctly, it should have the permissions inside (as metadata).
I verified it against those: https://github.com/tsuru/rpaas-operator/releases/tag/v0.33.1

Do you think we should override it, just in case?

Reply:
I think it's worth double check it. Otherwise a ~wild~ failed to exec plugin might apper.

Reply:
@nettoclaudio let me know what you think: https://github.com/tsuru/tsuru-client/pull/191/commits/86f18637cc66081c17cccef49a9110a2a2da1ac6

I cannot unit-test this, and I'm not reaaally sure what happens on windows. 
But it's non-destructive, so in the worst case, it will do nothing.

Reply:
Perfect!